Feature: Support VP230 for Huawei AC charger

schlimmchen opened the pull request:
There is a lot more in this PR under the hood, all concerning the Huawei grid charger, especially restructuring into namespace, splitting the driver and the controller, removing code duplication (web API JSON creation, (de)serialization of settings), etc.

AndreasBoehm commented on src/Configuration.cpp:
```diff
@@ -166,6 +167,23 @@ void ConfigurationClass::serializePowerLimiterConfig(PowerLimiterConfig const& s
+    target["auto_power_enabled"] = source.Auto_Power_Enabled;
+    target["auto_power_batterysoc_limits_enabled"] = source.Auto_Power_BatterySoC_Limits_Enabled;
+    target["emergency_charge_enabled"] = source.Emergency_Charge_Enabled;
+    target["voltage_limit"] = roundedFloat(source.Auto_Power_Voltage_Limit);
+    target["enable_voltage_limit"] = roundedFloat(source.Auto_Power_Enable_Voltage_Limit);
```
`roundedFloat` wasn't used for those two values before. Whats the reason that we use it now?

schlimmchen replied:
The rounding issue is present with all floats that are transported through ArduinoJson. The fact that roundedFloat() (or something similar) was not used, was a bug.

There was rounding before, but only in the web API implementation:
https://github.com/hoylabs/OpenDTU-OnBattery/blob/3f297f0df67beba191a3500e7b65c020f6e7fb42/src/WebApi_Huawei.cpp#L164-L165

Since we now use serialize...() for both writing the config and serving the web API, we use roundedFloat() here.

AndreasBoehm commented on src/Configuration.cpp:
```diff
@@ -477,6 +484,23 @@ void ConfigurationClass::deserializePowerLimiterConfig(JsonObject const& source,
+    target.Enabled = source["enabled"] | HUAWEI_ENABLED;
+    target.VerboseLogging = source["verbose_logging"] | VERBOSE_LOGGING;
+    target.HardwareInterface = source["hardware_interface"] | GridChargerHardwareInterface::MCP2515;
+    target.CAN_Controller_Frequency = source["can_controller_frequency"] | HUAWEI_CAN_CONTROLLER_FREQUENCY;
```
should we include **MCP2515** in the name of this config field and default value to better indicate that its only required when this hw interface is used?

schlimmchen replied:
Hm... The name already includes "CAN Controller", and the MCP2515 is a CAN controller, while the VP230 is a CAN transceiver. I don't think it's important enough to warrant a change and config upgrade. If you think we should do that, I can do it anyways. Is this the only config in question?

AndreasBoehm replied:
Makes sense, lets keep it as it is.

AndreasBoehm commented on src/main.cpp:
```diff
@@ -187,14 +187,8 @@ void setup()
-    } else {
-        MessageOutput.println("Invalid pin config");
-    }
+    MessageOutput.println("Initialize Huawei AC charger interface...");
```
Should we move this log to `HuaweiCan` to cleanup the main.cpp a bit? Would also align with other classes.

schlimmchen replied:
Yes, absolutely. :+1:

AndreasBoehm commented on webapp/src/locales/de.json:
```diff
@@ -312,7 +312,8 @@
         "Task_asynctcp": "Async TCP",
         "Task_mqttclient": "MQTT Client",
-        "Task_huaweican0": "AC Ladegerät CAN",
+        "Task_huaweihwifc": "Netzladegerät Hardwareschnittstelle",
```
`Hardwareschnittstelle` verwirrt mich etwas. Ist die Hardwareschnittstelle nicht entweder eine extra controller oder twai?

schlimmchen replied:
Ich meine hier damit den Task, der die Hardwareschnittstelle pollt. Also ja, entweder bespielt der den TWAI oder MCP2515. Wir können auch einfach "Netzladegerät" schreiben? Der weitere Task "huaweitwai" ist ein weiterer Task, der neben "huaweihwifc" läuft, aber nur wenn TWAI ausgewählt ist.

AndreasBoehm replied:
jetzt hab ichs verstanden, danke. ich denke weniger ist mehr: `Netzladegerät`

AndreasBoehm commented on webapp/src/types/HuaweiDataStatus.ts:
```diff
@@ -14,5 +14,4 @@ export interface Huawei {
     max_output_current: ValueObject;
     output_power: ValueObject;
     output_temp: ValueObject;
-    amp_hour: ValueObject;
```
I assume that you removed this because its unused, right?

schlimmchen replied:
Correct, I stumbled upon this and found that it is no used, so I cleaned it up. Objections?

AndreasBoehm replied:
no objections

AndreasBoehm commented on src/gridcharger/huawei/Controller.cpp:
```diff
@@ -0,0 +1,378 @@
+    { \
+        auto oDataPoint = _dataPoints.getDataPointFor<Label::l>(); \
+        if (oDataPoint) { \
+            root[n]["v"] = *_dataPoints.get<Label::l>(); \
```
Is there a way to put  `type` inside `DataPoint` to simplify its usage?
```
oDataPoint->getValue();
```

schlimmchen replied:
Yeah, this bugs me, too, believe me. However, we cannot make the DataPoint class a template, as the std::unordered_map in DataPointContainers can only hold elements of the same type. So the DataPoint is not aware of the type its std::variant is actually holding. As we can't determine the correct data type the "getValue()" function will return inside DataPoint, we have the get<>() inside DataPointContainer. I will address this when rethinking the DataPoint concept using code generation.

AndreasBoehm commented on src/WebApi_Huawei.cpp:
```diff
@@ -203,59 +167,12 @@ void WebApiHuaweiClass::onAdminPost(AsyncWebServerRequest* request)
     {
```
Do you want to extend the check if values are missing with the hwinterface?

schlimmchen replied:
I don't want to check anything in the web API POST implementation. As the API is designed for the Web UI, and since the Web UI works as we program it, I don't want to check the POST content. And if we do, we'll have to do it everywhere. Currently, the checks are very sloppy anyways. Also: Since we now use the deserialize...() funciton, we will put in default values for missing values, anyways. So... I think my answer is "no, I dont want to extend the checks". Okay with you?

AndreasBoehm replied:
makes sense

AndreasBoehm commented on src/gridcharger/huawei/Controller.cpp:
```diff
@@ -0,0 +1,378 @@
+
+void Controller::setParameter(float val, HardwareInterface::Setting setting)
+{
+    if (_mode == HUAWEI_MODE_AUTO_INT) { return; }
```
Not sure if i misunderstood this operation mode but shouldn't we still be able to set offline voltage and offline current?

schlimmchen replied:
I did not question this. Yes, you are right, one should at least be able to set offline limits. However, I think that one should be able to set all settings despite the mode, e.g., using the web UI or MQTT. If the Controller ist in auto mode, it shall correct the setting in the next loop.

schlimmchen replied:
To avoid a regression, I am going to explicitly allow setting the offline values in any case, but keep the check for the mode in place otherwise.

AndreasBoehm commented on src/WebApi_Huawei.cpp:
```diff
@@ -98,6 +71,8 @@ void WebApiHuaweiClass::onPost(AsyncWebServerRequest* request)
+
     if (root["voltage_valid"].is<bool>()) {
         if (root["voltage_valid"].as<bool>()) {
             if (root["voltage"].as<float>() < minimal_voltage || root["voltage"].as<float>() > 58) {
```
should we replace `58` with `HUAWEI_MINIMAL_OFFLINE_VOLTAGE`?

schlimmchen replied:
Well, this checks for a reasonable upper value of the voltage setting. Since this PSU can output 58.5V at most, I guess 58V makes sense. Let's keep it like this.

AndreasBoehm commented on src/gridcharger/huawei/Controller.cpp:
```diff
@@ -0,0 +1,378 @@
+
+        // Re-enable automatic power control if the output voltage has dropped below threshold
+        if (oOutputVoltage && *oOutputVoltage < config.Huawei.Auto_Power_Enable_Voltage_Limit ) {
+            _autoPowerEnabledCounter = 10;
```
Random number?

schlimmchen replied:
I'll tell you once I understand the Controller :man_shrugging: Until then, let's not touch it. Or ask @MalteSchm? Is there a particular reason to initialize this counter to 10? What does it do?

MalteSchm replied:
Hey and happy new year.
The general idea is that auto power is disabled when the output power drops below the minimum configured power limit. But this condition is also true when starting the PSU. To prevent that we disable the PSU right away I implemented this mechanism. Here a counter is initialized with 10 and decremented every time the output power is below the configured limit. In essence this introduces some latency shutting down the PSU. 

I believe a used this approach over a timestamp based version because I thought about cases where the conditions change so that the output power varies. But the values are arbitrary chosen and the intervals in which this code is run have also changed. So there may not be so much value in this approach now.

schlimmchen commented on webapp/src/locales/fr.json:
```diff
@@ -915,6 +915,9 @@
         "Configuration": "AC Charger Configuration",
```
Yeah, but I don't care. I can't translate those tokens, and the English ones are used as fallback anyways.

schlimmchen commented on src/gridcharger/huawei/Controller.cpp:
```diff
@@ -0,0 +1,378 @@
+        case GridChargerHardwareInterface::MCP2515:
+            _upHardwareInterface = std::make_unique<MCP2515>();
+            break;
+        case GridChargerHardwareInterface::VP230:
+            _upHardwareInterface = std::make_unique<TWAI>();
```
Yes. :+1:

schlimmchen commented on include/gridcharger/huawei/HardwareInterface.h:
```diff
@@ -0,0 +1,73 @@
+
+    std::unique_ptr<DataPointContainer> getCurrentData() { return std::move(_upDataCurrent); }
+
+    static uint32_t constexpr _dataRequestIntervalMillis = 2500;
```
Nope. #defines are global, or at least visible in the respective compilation unit. And possibly interfere at places one does not want them to be applied. Bad. Ugly. This definition, however, is a modern equivalent. It's a member variable, hence the underscore. It is public as the Controller uses this value to calculate a backoff. I cannot explain why... Hm, you are right, it should not have an underscore as it is public. Should also have a capital 'D' I guess?

AndreasBoehm replied:
```suggestion
    static uint32_t constexpr DataRequestIntervalMillis = 2500;
```

Sounds alright

AndreasBoehm commented on src/gridcharger/huawei/MCP2515.cpp:
```diff
@@ -0,0 +1,141 @@
+    MessageOutput.printf("[Huawei::MCP2515] clk = %d, miso = %d, mosi = %d, cs = %d, irq = %d\r\n",
+            pin.huawei_clk, pin.huawei_miso, pin.huawei_mosi, pin.huawei_cs, pin.huawei_irq);
+
+    if (pin.huawei_clk < 0 || pin.huawei_miso < 0 || pin.huawei_mosi < 0 || pin.huawei_cs < 0) {
```
Isn't `pin.huawei_irq` required as well?

schlimmchen replied:
Yes, indeed it is **not** optional. Good catch! Oh wait... This might be a breaking change. I guess the previous code would have worked with the IRQ pin not connected or defined at all. The task previously polled the MCP2515 IRQ pin with a crazy frequency (as often as possible), and `digitalRead()` might have returned `false` for pin `-1`, in which case the MCP2515 was polled regardless. So except for even more overhead because of unsuccessful polling, the CAN messages might have still arrived...

Meh... I am leaning strongly towards the breaking change and my new, event-driven concept, rather than throwing that out of the window to support something that I think should have never worked before. I think we can put a disclaimer in the Release Notes and expect people to connect and define te IRQ pin, if they actually have not, which I guess are very, very few? The wiki makes no mention of the wiring at all, but the example pin mapping has the IRQ pin set. So... Make it a Breaking Change and be done with it?

Of course, I will update the if-statement to also make the IRQ pin required.

AndreasBoehm replied:
Yes, lets make it a breaking change 👍

schlimmchen commented:
For transparency, this is what I did:
* Rebased onto current `development`
* Applied `yarn prettier` and fixed up the respective commit "Feature: huawei grid charger: support serial CAN controller"
* Used Andreas' suggestion to rename `_dataRequestIntervalMillis` and fixed up commit "huawei grid charger: merge common code in base class" (the commit I added based on his suggestion using the Github UI was incomplete as it only renamed the variable in the header, but not its uses).

drahdiwaberl commented:
oh. hope i can give it a try soon 🤗

gitisgreat2023 commented:
@schlimmchen I would be happy to test as well. I have the Christian's Pro PCB board. With the current solution using a MCP2515 6 pins are used ( power_pin, IRQ, MISO, MOSI, CS and CLK). With support for the VP230, how would it hardware like look like? I still need an opto coupler for slot detect, right?

schlimmchen commented:
Unless you have a CAN-based battery provider, you can just use this driver. It is part of the recent release (probably even in the one before...). Select it in the AC charger settings and adjust the pin mapping (define rx and tx that are routed to the VP230, should be rx=18 and tx=7). Then connect the Huawei according to the pinout of the RJ45. Pin 5 is CANL, Pin 4 is CANH.

> I still need an opto coupler for slot detect, right?

Yes, controlled by the "power" in the pin mapping.

gitisgreat2023 commented:
> Unless you have a CAN-based battery provider, you can just use this driver. It is part of the recent release (probably even in the one before...). Select it in the AC charger settings and adjust the pin mapping (define rx and tx that are routed to the VP230, should be rx=18 and tx=7). Then connect the Huawei according to the pinout of the RJ45. Pin 5 is CANL, Pin 4 is CANH.
> 
> > I still need an opto coupler for slot detect, right?
> 
> Yes, controlled by the "power" in the pin mapping.

So make sure I understand it correctly... for I moment I got confused. That SN65HVD230 chip is a RS485 transceiver right? I have at the moment a JK BMS connected using the RS485. Additionally I can use then the Huawei CAN? That chip SN65HVD230 which on the PCB Pro board is basically supporting RS485, it supports two devices simultaneously? Or that chip can do RS485 and CAN as well? 
I'm just confused why a Huawei CAN signal can be processed but a CAN battery not. Probably a stupid question but I don't get it.

schlimmchen commented:
I would like to "just" share the schematics with you, but that's @swingstate 's call to make, so I won't without his explicit permission. Let me instead explain:

The RJ45 port allows to connect something over RS485 and something else, possibly simultaneously, over CAN. Pins 1 and 2 of the RJ45 jack are connected to an RS485 transceiver. Pins 4 and 5 are connected to a CAN transceiver. Pin 6 is GND.

Does that make sense?

swingstate commented:
The board uses 2 transceiver, 1 for CAN (SN65...) and 1 for RS485 (RENESAS ISL317...). To confirm what schlimmchen posted - PINs 1 & 2 are for the RS485, 4 & 5 for CAN. 6 for GND.  
Technically you could use both transceiver, however I didn't test that nor do I know if the software supports that.

schlimmchen commented:
> nor do I know if the software supports that.

Sure, why wouldn't it? The ESP32 I/Os are mutually exclusive, so both transceivers can be used at the same time.

gitisgreat2023 commented:
@swingstate 
> The board uses 2 transceiver, 1 for CAN (SN65...) and 1 for RS485 (RENESAS ISL317...). To confirm what schlimmchen posted - PINs 1 & 2 are for the RS485, 4 & 5 for CAN. 6 for GND. Technically you could use both transceiver, however I didn't test that nor do I know if the software supports that.

In total I have to physically connect four wires from the Pro PCB to the Huawei & Opto coupler right?
CANH and CANL: two wires from Huawei to the LAN connector:
Define rx and tx that are routed to the VP230, should be rx=18 and tx=7. Then connect the Huawei according to the pinout of the RJ45. Pin 5 is CANL, Pin 4 is CANH.
Then +5V and the output of power pin to the input of the opto coupler . Here I have a question: I can use one of the four GPIO 39-42 available pins right for the power pin?
Then I connect the minus of the Huawei and slot detect wire to the output of the opto coupler as well.

Its okay to use 5V or should it be 3.3V for the opto coupler? I don't damage the board by using  with 5V? (Too high current drawn by the LED?)

gitisgreat2023 commented:
@schlimmchen 
I just tested it:
I configured like following:
`"huawei": {
        "rx": 18,
        "tx": 7,
        "power": 39        
        }`

![image](https://github.com/user-attachments/assets/af46dc02-50c4-4c56-9fa2-6c1818242e19)

Then  I log while saving the config. It gives the following error:
`09:05:09.480 > [Huawei::TWAI] rx = -1, tx = -1
09:05:09.529 > [Huawei::TWAI] invalid pin config
09:05:09.574 > [Huawei::Controller] Error initializing hardware interface
09:05:09.624 > [Huawei::TWAI] failed to stop driver`

schlimmchen commented:
Well, you did not apply the pin mapping correctly. Otherwise the Huawei::TWAI wouldn't read rx and tx to be -1.

schlimmchen commented:
> Irgendwie scheint was schief zu gehen bei den neuen tx rx Huawei pinzuordnungen...

Ne, sorry, aber das halte ich für ausgeschlossen.

Was steht denn in der UI im device manager? Hast du auch wirklich das richtige Profil ausgewählt und hast du gespeichert und dann neu gestartet?

gitisgreat2023 commented:
Owh man... Montagmorgen... neuen Profile nicht ausgewählt... :-(
Danke!
Ich hole mir noch mal einen zweiten Kaffee...  ;-)

gitisgreat2023 commented:
@schlimmchen 
Es ist nicht mein Tag.... was bedeutet das hier?!
11:51:03.067 > [Huawei::HwIfc] Failed to send data request

schlimmchen commented:
Dass das Huawei deine CAN Nachricht nicht ge-ACK't hat, heißt das.

* Hat das Huawei Strom?
* Hast du CANH/CANL richtig angeschlossen?
* Weißt du, dass das Netzteil funktionstüchtig ist?

gitisgreat2023 commented:
-Untere von Huawei -> CANH an blauer RJ45. Geht was kaputt wenn ich H und L mal tausche? Kann ich das mittels pin mapping ( 18 und 7 vertauschen bei rx und tx) oder muss ich es neu soldieren?
-LED Huawei: grün und rot brenen constant, orange blinkt
-Ja, vor einer halbe Stunde funktionierte er noch am anderen board.

Ich habe den powerpin noch nicht angeschlossen aber CAN funktioniert ja auch ohe, richtig?

gitisgreat2023 commented:
Also, habe noch zwei checks gemacht:
Direkt am RJ45 Stecker und Stecker die an dem Huawei geht gemessen, CANH (blauer RJ45) und CANL (blau/weiss) 0.6 Ohm).
Dann den Huawei an dem alten board angesteckt, *ohne* den slot detect stecker (um zu checken dass CAN ohne slot detect angeschlossen geht): ja kein Thema, im GUI taucht er direkt auf.

Mir gehen die Ideeen aus. Haben den Huawei am alten board gelassen und er pumpt jetzt fleißig wieder Energie in die Batterie.

@schlimmchen hat irgendjemand mal diesen VP230 mit dem Huawei hardwäremäßig am laufen gehabt?

schlimmchen commented:
> hat irgendjemand mal diesen VP230 mit dem Huawei hardware und softwaremäßig am laufen gehabt?

Man hört ja oft, dass es keine blöden Fragen gibt. Aber sonderlich qualifiziert ist diese ja nicht, oder? Gewiss habe ich das nicht blind implementiert und nie getestet...

Bei mir läuft das wunderbar an einem OpenDTU Fusion mit CAN/Iso Shield. Ich kann es heute Abend mal an meinem OnBattery Pro versuchen.

> Untere pin vom Huawei -> CANH an blauer RJ45.

Blau... Damit ist nicht festgelegt, welchen Pin diese Adern am Ende berührt. Aber wahrscheinlich richtig, das Kabel wird sich wohl an einen der gängigen Standards halten.

> Geht was kaputt wenn ich H und L mal tausche?

Soweit ich (aus praktischer, ungewollter Erfahrung) glaube zu wissen: Nein.

> Kann ich das mittels pin mapping ( 18 und 7 vertauschen bei rx und tx) oder muss ich es neu löten?

Nein, das geht nicht über die Pin Mapping, denn die Pins am Transceiver ICs tauschen nicht einfach ihre Rolle, wenn der ESP32 sie anders zu interpretieren versucht.

gitisgreat2023 commented:
> > hat irgendjemand mal diesen VP230 mit dem Huawei hardware und softwaremäßig am laufen gehabt?
> 
> Man hört ja oft, dass es keine blöden Fragen gibt. Aber sonderlich qualifiziert ist diese ja nicht, oder? Gewiss habe ich das nicht blind implementiert und nie getestet...

:-D Mea culpa. Das war tätsichlich eine blöde Frage... klar, ist getestet und funktioniert an einem Fusion board. Nicht an dem Pro Board. Assumption is the mother of all screw-ups... wusste nicht dass genauso ein chip auch im Fusion drin ist. Alles klar, damit funktioniert es.

Ja, sehr gerne dass du es mit dem Pro versuchst.
Ich habe auch beim Pro es versucht ohne dass das JK BMS aktiv war (nur im GUI ausgeschaltet, das pinmapping war noch aktiv!). 
> 
> Bei mir läuft das wunderbar an einem OpenDTU Fusion mit CAN/Iso Shield. Ich kann es heute Abend mal an meinem OnBattery Pro versuchen.
> 
> > Untere pin vom Huawei -> CANH an blauer RJ45.
> 
> Blau... Damit ist nicht festgelegt, welchen Pin diese Adern am Ende berührt. Aber wahrscheinlich richtig, das Kabel wird sich wohl an einen der gängigen Standards halten.

My bad... nicht nur die Farbe sondern auch die Nummer (der clip weisst von mir weg) 4 hätte ich erwähnen sollen...

> 
> > Geht was kaputt wenn ich H und L mal tausche?
> 
> Soweit ich (aus praktischer, ungewollter Erfahrung) glaube zu wissen: Nein.

Okay... ich mache es doch noch mal, nicht dass beim hw H und L doch anders sind...

> 
> > Kann ich das mittels pin mapping ( 18 und 7 vertauschen bei rx und tx) oder muss ich es neu löten?
> 
> Nein, das geht nicht über die Pin Mapping, denn die Pins am Transceiver ICs tauschen nicht einfach ihre Rolle, wenn der ESP32 sie anders zu interpretieren versucht.

Ich verstehe das Konzept pin mapping dann nicht. Wenn ich "RX" 18 und "TX" 7 habe, dann interpretiert er doch GPIO 18 als empfangen und 7 als senden? Ich dacht die Flexibilität kommt daher dass er die inputs "mappen" kann.

Was wird dann genau gemapped? (hoffentlich zählt der hier nicht als unqualifiziert ;-))

@swingstate da das Eyes Emoji von dir kommt... weißt du ob jemand erfolgreich den Huawei am Pro Board mittels VP230 zu laufen bekommen hat?

swingstate commented:
Nope. Steht auf meiner todo liste.

gitisgreat2023 commented:
Alles klar. Ich habe mal CANH und CANL hardwaremäßig vertauscht. Macht keinen Unterschied.
Er hängt wieder am alten Board. Ich bin gespannt ob es bei @schlimmchen klappt... hoffentlich liegt der Fehler irgendwie bei mir.

schlimmchen commented:
> Ich verstehe das Konzept pin mapping dann nicht. Wenn ich "RX" 18 und "TX" 7 habe, dann interpretiert er doch GPIO 18 als empfangen und 7 als senden? Ich dacht die Flexibilität kommt daher dass er die inputs "mappen" kann.

Du kannst jeden (?) GPIO eines ESP32 intern beliebig verbinden mit einer Funktion, wie dem TWAI. Man kann also jeden GPIO zu jenem Pin erklären, der vom TWAI als Ausgang für Daten verwendet wird, und jeden GPIO zu jenem Pin, von dem Daten gelesen werden. Das passiert intern im ESP32.

Wenn du im Pin Mapping definierst, dass ein bestimmter GPIO mit TWAI RX verbunden sein soll im ESP32, dann muss der GPIO aber auch phyisch per Kupfer mit dem RX Pin des Transceivers verbunden sein. Wenn du dieses RX/TX Paar im Pin Mapping vertauschst, dann erreichst du nur, dass beide Komponenten auf dem jeweils einen Signal (Kupferleitung) beide sprechen (und dabei kann was kaputt gehen) und auf dem jeweils anderen Signal (Kupferleitung) beide hören.

Hinzu kommt noch der Logik-Fehler, dass CAN ein Bus ist und CANH und CANL ein differenzielles Signal ist. Es ist nicht eins "RX" und das andere "TX", sondern das eine ist das invertierte Signal des anderen. Daher bringt das Tauschen im Pin Mapping erst recht nichts.

Puh... So audwändig aufgeschrieben und trotzdem glaube ich, dass mans nicht versteht. Ich fürchte das nächste Mal hacke ich das ganz frech ein die KI ein und lass es mir raus... So in etwa :grin: 

## AI output

**Warum kann man RX und TX beim CAN-Bus nicht einfach im Pin-Mapping tauschen?**  

Das Pin-Mapping des ESP32 erlaubt es, fast jede Hardware-Funktion auf fast jeden GPIO zu legen. Das bedeutet aber nur, dass z. B. die TWAI-Einheit des ESP32 entscheiden kann, an welchen Pins sie Daten senden (TX) und empfangen (RX) soll.  

Der CAN-Bus selbst funktioniert jedoch nach einem anderen Prinzip:  
- Der ESP32 sendet digitale Signale über den TX-Pin an den CAN-Transceiver.  
- Der Transceiver übersetzt diese Signale auf die physikalischen CANH- und CANL-Leitungen.  
- Die empfangenen Signale werden vom Transceiver wieder in digitale Pegel gewandelt und über den RX-Pin zurück an den ESP32 gegeben.  

Wenn CANH und CANL vertauscht sind, dann bekommt der Transceiver falsche Differenzspannungen. Dadurch wird die Buskommunikation gestört oder gar nicht funktionieren. Das reine Tauschen von RX und TX in der Pin-Zuordnung des ESP32 hilft hier nicht, weil der Fehler vor dem Mikrocontroller, nämlich auf der physikalischen Leitung, passiert.  

**Lösung:** Die einzige korrekte Lösung ist, die Verdrahtung zu überprüfen und CANH/CANL richtig anzuschließen. Alternativ könnte ein Transceiver-Modul mit invertierten Signalen genutzt werden, aber das ist normalerweise nicht nötig, wenn die Verkabelung korrekt ist.

schlimmchen commented:
Klappt wunderbar.

![image](https://github.com/user-attachments/assets/1f30fe27-5ece-42cf-8457-048d0a323322)

* Pin 5 des RJ45 Steckers ist CANL, mit einem T-568B codierten Kabel ist das blau/weiß
* Pin 4 des RJ45 Steckers ist CANH, mit einem T-568B codierten Kabel ist das blau
* Die Kommunikation funktioniert bei mir nur mit CAN Abschlusswiderstand, daher muss der Jumper an H5 gesetzt sein und zwar wie hier gezeigt:

![image](https://github.com/user-attachments/assets/e8215ba3-58de-4f40-8608-fc716db7711c)

Und wie bereits behauptet ist die RS485 Schnittstelle (hier zum JK BMS) gleichzeitig nutzbar.

@swingstate Warum ist der Abschlusswiderstand optional? Ich glaube das macht mehr Probleme, als es löst, denkst du nicht? Markus hat sich dafür entschieden, den fest einzulöten auf dem CAN/Iso Shield.

gitisgreat2023 commented:
@schlimmchen Du bist mein Held heute! Der Abschlusswiederstand hat gefehlt! Es läuft! Vielen Dank!!

swingstate commented:
> @swingstate Warum ist der Abschlusswiderstand optional? Ich glaube das macht mehr Probleme, als es löst, denkst du nicht? Markus hat sich dafür entschieden, den fest einzulöten auf dem CAN/Iso Shield.

Es gab hier mal Berichte, dass es bei aktivem Widerstand Schwierigkeiten mit der CAN Verbindung gab. Daher optional bzw. in neuen Revisionen schaltbar und dann standardmäßig "ein"-geschaltet.
